File: tls.erl
-module(tls).

% TLS Server
-export([server/0]).
% Clients that can connect to the local TLS server (defined in this module too)
-export([client_certificate_valid/0, client_certificate_invalid/0]).
% Client that can connect to the ATS TLS server:
% - you need to supply password to client key somehow)
% - if you'd like to have server side verification failing, then replace client fingerprint in the ATS config
-export([iwf_client/0]).

-define(LOCAL_SERVER_ADDRESS, "localhost").
-define(IWF_SERVER_ADDRESS, "10.140.192.25").

-define(PORT, 9000).
-define(CONNECT_TIMEOUT, infinity).

-define(LOCAL_CLIENT_CERT, "/home/amj018/tmp/otp_certs/client/cert.pem").
-define(LOCAL_CLIENT_KEY, "/home/amj018/tmp/otp_certs/client/key.pem").
-define(IWF_CLIENT_CERT, "/usr/app/iwf/certs/iwf.cert.pem").
-define(IWF_CLIENT_KEY, "/usr/app/iwf/certs/private/iwf.key.pem").
%-define(CLIENT_PASSWORD, "").

-define(LOCAL_SERVER_CERT, "/home/amj018/tmp/otp_certs/server/cert.pem").
-define(LOCAL_SERVER_KEY, "/home/amj018/tmp/otp_certs/server/key.pem").

-define(LOCAL_INVALID_CERT, "/home/amj018/tls-excercise/certs/invalid.pem").
-define(LOCAL_INVALID_KEY, "/home/amj018/tls-excercise/certs/invalid.key").

-define(LOCAL_CACERTS, "/home/amj018/tmp/otp_certs/cacerts.pem").
-define(ATS_CACERTS, "/usr/app/iwf/certs/ats.cert.pem").

%
% Exported functions
%

server() ->
    server(?LOCAL_SERVER_CERT, ?LOCAL_SERVER_KEY, ?LOCAL_CACERTS).

client_certificate_valid() ->
    client(?LOCAL_CLIENT_CERT, ?LOCAL_CLIENT_KEY, ?LOCAL_CACERTS).

client_certificate_invalid() ->
    client(?LOCAL_INVALID_CERT, ?LOCAL_INVALID_KEY,  ?LOCAL_CACERTS).

iwf_client() ->
    client(?IWF_CLIENT_CERT, ?IWF_CLIENT_KEY, ?ATS_CACERTS).

%
% Internal functions
%

client(ClientCert, ClientKey, Cacerts) ->
    ssl:start(),
    TlsOptions = [
                  {mode, binary},
                  {active, false},
                  {tos, 224},
                  {versions, ['tlsv1.2']},
                  %{ciphers, ???},
                  {certfile, ClientCert},
                  {keyfile, ClientKey},
                  %{password, ?CLIENT_PASSWORD},
                  {certs_keys, [#{certfile => ClientCert, keyfile => ClientKey}]},
                  {verify_fun, {fun(_,{bad_cert, _} = Reason, _) ->
                                        %io:format("line: ~p~n", [?LINE]),
                                        {fail, Reason};
                                   (_,{extension, _}, UserState) ->
                                        %io:format("line: ~p~n", [?LINE]),
                                        {unknown, UserState};
                                   (_, valid, UserState) ->
                                        %io:format("line: ~p~n", [?LINE]),
                                        {valid, UserState};
                                   (_, valid_peer, UserState) ->
                                        %io:format("line: ~p~n", [?LINE]),
                                        {valid, UserState}
                                end, []}
                  },
                  {keepalive, true},
                  {nodelay, true}, %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

                  {cacertfile, Cacerts},
                  {verify, verify_peer},
                  {server_name_indication, disable}
                 ],

    io:format("Before ssl:connect~n"),
    {ok, Socket} = ssl:connect(?LOCAL_SERVER_ADDRESS, ?PORT, TlsOptions, ?CONNECT_TIMEOUT),
    io:format("After ssl:connect~n"),
    {ok, Data} = ssl:recv(Socket, 0, infinity),
    io:format("Received: ~p~n", [Data]),
    ok = ssl:send(Socket, Data),
    io:format("Send    : ~p~n", [Data]),
    timer:sleep(5 * 1000),
    ssl:close(Socket),
    ssl:stop().

server(ServerCert, ServerKey, Cacerts) ->
    ssl:start(),
    TlsOptions = [{certs_keys, [#{certfile => ServerCert,
                                  keyfile => ServerKey}]},
                  {reuseaddr, true},
                  {versions, ['tlsv1.2']},
                  {cacertfile, Cacerts},

                  {verify, verify_peer},
                  {verify_fun, {fun(_,{bad_cert, _} = Reason, _) ->
                                        %io:format("line: ~p~n", [?LINE]),
                                        {fail, Reason};
                                   (_,{extension, _}, UserState) ->
                                        %io:format("line: ~p~n", [?LINE]),
                                        {unknown, UserState};
                                   (_, valid, UserState) ->
                                        %io:format("line: ~p~n", [?LINE]),
                                        {valid, UserState};
                                   (_, valid_peer, UserState) ->
                                        %io:format("line: ~p~n", [?LINE]),
                                        {valid, UserState}
                                end, []}
                  }
                 ],

    HandshakeTlsOptions = [
                           [{verify, verify_peer},
                            {fail_if_no_peer_cert, true},
                            {cacertfile, Cacerts},
                            {certs_keys, [#{certfile => ServerCert, keyfile => ServerKey}]}]
                          ],

    {ok, ListenSocket} = ssl:listen(?PORT, TlsOptions),
    listen(ListenSocket, HandshakeTlsOptions).

listen(ListenSocket, HandshakeTlsOptions) ->
    io:format("Server ready~n", []),
    {ok, TLSTransportSocket} = ssl:transport_accept(ListenSocket),
    {ok, Socket} = ssl:handshake(TLSTransportSocket, HandshakeTlsOptions),
    Data = "server ping",
    ok = ssl:send(Socket, Data),
    io:format("Server send: ~p~n", [Data]),
    receive
        {ssl, Socket, Data} ->
            io:format("Server received: ~p~n", [Data])
    end,
    ssl:shutdown(Socket, read_write),
    ssl:close(Socket),
    timer:sleep(2 * 1000),
    listen(ListenSocket, HandshakeTlsOptions).

File: tls.py
#!/usr/bin/env python3

import socket, ssl, sys, os, time

PORT= 10021
SERVER_CERT="certs/server.pem"
SERVER_KEY="certs/server.key"

CLIENT_CERT="certs/client.pem"
CLIENT_KEY="certs/client.key"

INVALID_CERT="certs/invalid.pem"
INVALID_KEY="certs/invalid.key"

class Server:
    def __init__(self, mode, port, server_cert, server_key, client_cert = None):
        self.mode = mode
        self.port = port
        self.server_cert = server_cert
        self.server_key = server_key
        self.client_cert = client_cert

    def start(self):
        context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)

        context.load_default_certs()
        context.load_cert_chain(certfile=self.server_cert, keyfile=self.server_key)
        if self.client_cert != None:
            context.verify_mode = ssl.CERT_REQUIRED
            context.load_verify_locations(CLIENT_CERT)

        bindsocket = socket.socket()
        bindsocket.bind(('localhost', self.port))
        bindsocket.listen()

        while True:
            try:
                print("[{}] Waiting for new client".format(self.mode))
                newsocket, fromaddr = bindsocket.accept()
                try:
                    connstream = context.wrap_socket(newsocket, server_side=True)
                    print("Client connected: {}".format(repr(connstream)))
                    self.deal_with_client(connstream)
                except ssl.SSLError as e:
                    print("Unsuccessfull connection attempt: {}".format(e))
            except KeyboardInterrupt:
                print("KeyboardInterrupt")
                if 'connstream' in locals():
                    connstream.shutdown(socket.SHUT_RDWR)
                    connstream.close()
                sys.exit(0)

    def deal_with_client(self, connstream):
        while True:
            data = connstream.recv(1024)
            if data != b'':
                print("Received (p:{}): {}".format(self.port, repr(data)))
                connstream.send(data)
                time.sleep(1)
                print("Done with client")
                connstream.shutdown(socket.SHUT_RDWR)
                connstream.close()
                return

def client(mode, port, server_cert, client_cert = None, client_key = None):
    print("[{}]: Started".format(mode))
    context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH, cafile=server_cert)
    if client_cert != None:
        context.load_cert_chain(certfile=client_cert, keyfile=client_key)
    conn = context.wrap_socket(socket.socket(socket.AF_INET),
                               server_hostname="localhost")
    conn.connect(("localhost", PORT))
    to_send = b"ping"
    conn.send(to_send)
    print("Send    : {}".format(to_send))
    data = conn.recv(1024)
    print("Received: {}".format(data))
    conn.shutdown(socket.SHUT_RDWR)
    conn.close()

def usage():
    print("Usage: {} [mode]".format(os.path.basename(sys.argv[0])))
    print("  mode = [ client | iwf-client | client-with-wrong-own-certificate | client-with-wrong-server-certificate ]".format(os.path.basename(sys.argv[0])))
    print("  mode = [ server | ats-server ]".format(os.path.basename(sys.argv[0])))
    sys.exit(1)

if len(sys.argv) < 2:
    usage()

mode = os.path.basename(sys.argv[1])

if mode not in ["client", "iwf-client", "client-with-wrong-own-certificate", "client-with-wrong-server-certificate", "server", "ats-server"]:
    usage()

match mode:
    case "client":
        client(mode, PORT, SERVER_CERT)
    case "iwf-client": # This client can verify itself to the server
        client(mode, PORT, SERVER_CERT, CLIENT_CERT, CLIENT_KEY)
    case "client-with-wrong-own-certificate":
        client(mode, PORT, SERVER_CERT, INVALID_CERT, INVALID_KEY)
    case "client-with-wrong-server-certificate":
        client(mode, PORT, INVALID_CERT, CLIENT_CERT, CLIENT_KEY)
    case "server":
        Server(mode, PORT, SERVER_CERT, SERVER_KEY).start()
    case "ats-server": # This server will request certificate from client
        Server(mode, PORT, SERVER_CERT, SERVER_KEY, CLIENT_CERT).start()

